File: ios/runanywhere-sdks-latest/sdk/runanywhere-swift/Sources/RunAnywhere/Public/Extensions/VoiceAgent/RunAnywhere+VoiceAgent.swift
//
//  RunAnywhere+VoiceAgent.swift
//  RunAnywhere SDK
//
//  Public API for Voice Agent operations (full voice pipeline).
//  Calls C++ directly via CppBridge for all operations.
//  Events are emitted by C++ layer - no Swift event emissions needed.
//
//  Architecture:
//  - Voice agent uses SHARED handles from individual components (STT/LLM/TTS/VAD)
//  - Models are loaded via the canonical lifecycle (`RAModelLoadRequest`)
//  - Voice agent is purely an orchestrator for the full voice pipeline
//  - All events (including state changes) are emitted from C++
//

import CRACommons
import Foundation

// MARK: - Voice Agent Operations

public extension RunAnywhere {

    /// Initialize the voice agent with configuration.
    static func initializeVoiceAgent(_ config: RAVoiceAgentComposeConfig) async throws {
        guard isInitialized else {
            throw SDKException(code: .notInitialized, message: "SDK not initialized", category: .internal)
        }

        try await ensureServicesReady()
        _ = try await CppBridge.VoiceAgent.shared.getHandle()
        _ = try await CppBridge.VoiceAgent.shared.initialize(config)
    }

    /// Default Silero VAD model id seeded by every example app's catalog.
    /// Exposed so callers do not hard-code the string when invoking
    /// `ensureDefaultVAD(...)`.
    static var defaultVADModelID: String { "silero-vad" }

    /// Ensure a VAD model is loaded in the canonical lifecycle before a voice
    /// agent session starts. When no VAD model is currently registered for
    /// `.voiceActivityDetection`, attempts to load the catalogued default
    /// (`defaultVADModelID`, Silero) so the voice agent's speech-start /
    /// speech-end events fire. The energy-based fallback does not produce
    /// the lifecycle events the voice-agent orchestrator listens for, so
    /// without a VAD lifecycle load the session stays silent after init.
    ///
    /// Idempotent: returns `true` immediately when a VAD model is already
    /// loaded. Logs (but does not throw) when the optional auto-load fails;
    /// callers may inspect the return value to decide whether to surface a
    /// warning.
    ///
    /// - Parameter modelID: VAD model id to auto-load when none is current.
    ///   Defaults to `defaultVADModelID`.
    /// - Returns: `true` when a VAD model is loaded after the call;
    ///   `false` when no VAD model is loaded (auto-load failed or skipped).
    @discardableResult
    static func ensureDefaultVAD(modelID: String? = nil) async -> Bool {
        guard isInitialized else { return false }

        let snapshot = loadedModelSnapshot(category: .voiceActivityDetection)
        if snapshot.found && !snapshot.modelID.isEmpty {
            return true
        }

        let targetID = modelID ?? defaultVADModelID
        guard !targetID.isEmpty else { return false }

        SDKLogger.voiceAgent.info("Auto-loading default VAD '\(targetID)' for voice-agent session")

        var loadRequest = RAModelLoadRequest()
        loadRequest.modelID = targetID
        loadRequest.category = .voiceActivityDetection
        let result = await RunAnywhere.loadModel(loadRequest)
        if !result.success {
            SDKLogger.voiceAgent.warning(
                "Default VAD '\(targetID)' auto-load failed: \(result.errorMessage) — voice agent will use energy fallback"
            )
            return false
        }
        return true
    }

    /// Initialize the voice agent from currently-loaded STT / LLM / TTS models.
    ///
    /// Composes a `RAVoiceAgentComposeConfig` from the canonical model
    /// lifecycle (`RunAnywhere.currentModel(_:)`) snapshots for
    /// `.speechRecognition`, `.language`, and `.speechSynthesis`, then
    /// forwards to `initializeVoiceAgent(_:)`. Mirrors the Kotlin / Web SDKs'
    /// `initializeVoiceAgentWithLoadedModels()` API.
    ///
    /// When `ensureVAD` is `true` (default), the SDK guarantees that a VAD
    /// model is loaded into the canonical lifecycle before initialization
    /// runs via `ensureDefaultVAD(...)`. Without this the session would
    /// silently fall back to the energy-based detector and the C++ voice
    /// agent's speech-start / speech-end lifecycle events would not fire.
    /// Set to `false` only if the caller has already loaded an explicit VAD
    /// model (or knows the energy fallback is acceptable for the deployment).
    ///
    /// - Parameter ttsVoiceID: Optional explicit voice id to pass through to
    ///   the TTS engine. For multi-voice TTS engines (e.g., Sherpa-ONNX-TTS
    ///   with Piper multi-speaker models), the voice id selects which voice
    ///   within the loaded model to use and is semantically distinct from the
    ///   TTS model id. When `nil` (the default), the engine's default voice
    ///   is used — appropriate for single-voice models. Callers using
    ///   multi-voice engines should pass the desired voice id explicitly.
    ///   Never reuse the TTS model id here — model id ≠ voice id.
    /// - Parameter ensureVAD: Whether to auto-load the catalogued default VAD
    ///   when no `.voiceActivityDetection` model is loaded. Defaults to
    ///   `true`.
    /// - Throws: `SDKException(code: .notInitialized, message: ..., category: .internal)` if the SDK has
    ///           not completed Phase 1 initialization.
    /// - Throws: `SDKException(code: .modelNotLoaded, message: ..., category: .component)` if STT, LLM,
    ///           or TTS has no model loaded.
    static func initializeVoiceAgentWithLoadedModels(
        ttsVoiceID: String? = nil,
        ensureVAD: Bool = true
    ) async throws {
        guard isInitialized else {
            throw SDKException(code: .notInitialized, message: "SDK not initialized", category: .internal)
        }

        try await ensureServicesReady()

        if ensureVAD {
            _ = await ensureDefaultVAD()
        }

        // The C++ lifecycle service is the canonical source of truth for
        // "is this modality loaded"; the per-component CppBridge actor
        // mirrors are not updated by RunAnywhere.loadModel(_:). Query the
        // lifecycle directly, matching the iOS example app and the rest of
        // the public Swift surface (STT/TTS/VLM readiness checks).
        let sttSnap = loadedModelSnapshot(category: .speechRecognition)
        let llmSnap = loadedModelSnapshot(category: .language)
        let ttsSnap = loadedModelSnapshot(category: .speechSynthesis)

        var missing: [String] = []
        if !sttSnap.found || sttSnap.modelID.isEmpty { missing.append("STT") }
        if !llmSnap.found || llmSnap.modelID.isEmpty { missing.append("LLM") }
        if !ttsSnap.found || ttsSnap.modelID.isEmpty { missing.append("TTS") }
        guard missing.isEmpty else {
            throw SDKException(
                code: .modelNotLoaded,
                message: "Cannot initialize voice agent: Models not loaded: \(missing.joined(separator: ", "))",
                category: .component
            )
        }

        var config = RAVoiceAgentComposeConfig()
        config.sttModelID = sttSnap.modelID
        config.llmModelID = llmSnap.modelID
        // ttsVoiceID is the voice id *within* the loaded TTS model, NOT the
        // model id. For single-voice engines, leaving it unset lets the
        // engine pick its default voice. For multi-voice engines (Piper,
        // eSpeak-NG, Sherpa-ONNX-TTS multi-voice), the caller must supply
        // the desired voice id explicitly via the `ttsVoiceID` parameter.
        // Previous releases conflated `ttsSnap.modelID` with the voice id,
        // which produced an invalid voice selection for multi-voice models.
        if let voiceID = ttsVoiceID, !voiceID.isEmpty {
            config.ttsVoiceID = voiceID
        }

        _ = try await CppBridge.VoiceAgent.shared.getHandle()
        _ = try await CppBridge.VoiceAgent.shared.initialize(config)
    }

    /// Get the current voice-agent component states (per-component load status
    /// and the aggregate `ready` flag).
    ///
    /// Mirrors the Kotlin / Web SDKs' `getVoiceAgentComponentStates()` API.
    /// Thin forwarder over `CppBridge.VoiceAgent.shared.componentStatesProto()`.
    ///
    /// - Returns: Proto `RAVoiceAgentComponentStates` with per-component
    ///            `ComponentLifecycleState` and a computed `ready` flag.
    /// - Throws: `SDKException(code: .notInitialized, message: ..., category: .internal)` if the SDK has
    ///           not completed Phase 1 initialization.
    static func getVoiceAgentComponentStates() async throws -> RAVoiceAgentComponentStates {
        guard isInitialized else {
            throw SDKException(code: .notInitialized, message: "SDK not initialized", category: .internal)
        }

        try await ensureServicesReady()
        return try await CppBridge.VoiceAgent.shared.componentStates()
    }

    /// Process a complete voice turn through the proto C++ ABI.
    static func processVoiceTurn(_ audioData: Data) async throws -> RAVoiceAgentResult {
        guard isInitialized else {
            throw SDKException(code: .notInitialized, message: "SDK not initialized", category: .internal)
        }

        try await ensureServicesReady()

        guard await CppBridge.VoiceAgent.shared.isReady else {
            throw SDKException(code: .notInitialized, message: "Voice agent not ready", category: .component)
        }

        return try await CppBridge.VoiceAgent.shared.processVoiceTurnProto(audioData)
    }

    /// Open a stream of canonical `RAVoiceEvent` proto events for the active
    /// voice agent.
    ///
    /// While the stream is consumed, a platform mic driver captures audio,
    /// forwards raw frames via `rac_voice_agent_feed_audio_proto`.
    /// Events fan out through the handle callback registered by
    /// `VoiceAgentStreamAdapter`.
    ///
    /// Cancellation: breaking out of the consuming `for-await` loop (or
    /// cancelling the surrounding `Task`) tears down mic capture and the C
    /// callback via `rac_voice_agent_set_proto_callback(handle, nullptr, nullptr)`.
    static func streamVoiceAgent() -> AsyncStream<RAVoiceEvent> {
        AsyncStream { continuation in
            let task = Task {
                do {
                    guard isInitialized else {
                        continuation.finish()
                        return
                    }

                    try await ensureServicesReady()
                    let handle = try await CppBridge.VoiceAgent.shared.getHandle()

                    let micDriver = VoiceAgentMicDriver(handle: handle)
                    let micTask = Task {
                        do {
                            try await micDriver.run()
                        } catch is CancellationError {
                            // Expected when the consumer stops the session.
                        } catch {
                            SDKLogger.voiceAgent.error("Voice-agent mic driver stopped: \(error.localizedDescription)")
                        }
                    }

                    defer { micTask.cancel() }

                    let adapter = VoiceAgentStreamAdapter(handle: handle.rawValue)
                    for await event in adapter.stream() {
                        if Task.isCancelled { break }
                        continuation.yield(event)
                    }
                } catch {
                    SDKLogger.voiceAgent.error("Voice agent stream setup failed: \(error.localizedDescription)")
                }
                continuation.finish()
            }

            continuation.onTermination = { @Sendable _ in
                task.cancel()
            }
        }
    }

    /// Cleanup voice agent resources.
    static func cleanupVoiceAgent() async {
        await CppBridge.VoiceAgent.shared.cleanup()
    }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-react-native/packages/core/src/Public/Extensions/VoiceAgent/RunAnywhere+VoiceAgent.ts
/**
 * RunAnywhere+VoiceAgent.ts
 *
 * Voice agent (full VAD → STT → LLM → TTS) extension. All shapes come from
 * `@runanywhere/proto-ts/voice_agent_service` and
 * `@runanywhere/proto-ts/voice_events`; commons owns the pipeline.
 *
 * Mirrors `sdk/runanywhere-swift/Sources/RunAnywhere/Public/Extensions/VoiceAgent/RunAnywhere+VoiceAgent.swift`.
 */

import { requireNativeModule, isNativeModuleAvailable } from '../../../native';
import { SDKLogger } from '../../../Foundation/Logging/Logger/SDKLogger';
import { SDKException } from '../../../Foundation/Errors/SDKException';
import { ErrorCode as ErrorCodeProto } from '@runanywhere/proto-ts/errors';
import type {
  VoiceAgentResult,
} from '@runanywhere/proto-ts/voice_agent_service';
import {
  VoiceAgentComposeConfig,
  VoiceAgentResult as VoiceAgentResultMessage,
} from '@runanywhere/proto-ts/voice_agent_service';
import type { VoiceAgentComponentStates } from '@runanywhere/proto-ts/voice_events';
import {
  VoiceAgentComponentStates as VoiceAgentComponentStatesMessage,
} from '@runanywhere/proto-ts/voice_events';
import type { VoiceEvent } from '@runanywhere/proto-ts/voice_events';
import {
  CurrentModelRequest,
  ModelCategory,
  ModelLoadRequest,
} from '@runanywhere/proto-ts/model_types';
import { currentModel, loadModel } from '../Models/RunAnywhere+ModelLifecycle';
import { VoiceAgentStreamAdapter } from '../../../Adapters/VoiceAgentStreamAdapter';
import { VoiceAgentMicDriver } from '../../../Features/VoiceAgent/VoiceAgentMicDriver';
import { arrayBufferToBytes, bytesToArrayBuffer } from '../../../services/ProtoBytes';
import { ensureServicesReady } from '../../../Foundation/Initialization/ServicesReadyGuard';
import {
  isSDKInitialized,
  requireInitialized,
} from '../../../Foundation/Initialization/InitializedGuard';
import { encodeProtoMessage } from '../../../services/ProtoWire';

const logger = new SDKLogger('RunAnywhere.VoiceAgent');

function ensureNative() {
  if (!isNativeModuleAvailable()) {
    throw SDKException.nativeModuleUnavailable();
  }
  return requireNativeModule();
}

function audioToArrayBuffer(audioData: ArrayBuffer | Uint8Array): ArrayBuffer {
  if (audioData instanceof Uint8Array) {
    return bytesToArrayBuffer(audioData);
  }
  return audioData;
}

/**
 * Get voice agent component states.
 *
 * Matches Swift: `RunAnywhere.getVoiceAgentComponentStates()`.
 */
export async function getVoiceAgentComponentStates(): Promise<VoiceAgentComponentStates> {
  // Swift parity: guard isInitialized (RunAnywhere+VoiceAgent.swift:178-180).
  requireInitialized();
  const native = ensureNative();
  // Swift parity: RunAnywhere+VoiceAgent.swift:182 gates on ensureServicesReady.
  await ensureServicesReady();
  try {
    const bytes = await native.voiceAgentComponentStatesProto();
    return VoiceAgentComponentStatesMessage.decode(arrayBufferToBytes(bytes));
  } catch (error) {
    const msg = error instanceof Error ? error.message : String(error);
    logger.error(`Failed to get component states: ${msg}`);
    throw error;
  }
}

/** Initialize voice agent with the canonical proto compose config. Mirrors Swift's `initializeVoiceAgent(_ config:)`. */
export async function initializeVoiceAgent(
  config: ReturnType<typeof VoiceAgentComposeConfig.create>
): Promise<void> {
  // Swift parity: guard isInitialized (RunAnywhere+VoiceAgent.swift:25-27).
  requireInitialized();
  const native = ensureNative();
  // Swift parity: RunAnywhere+VoiceAgent.swift:29 gates on ensureServicesReady.
  await ensureServicesReady();
  try {
    logger.info('Initializing voice agent...');
    const bytes = await native.voiceAgentInitializeProto(
      encodeProtoMessage(config, VoiceAgentComposeConfig)
    );
    // Swift parity: initializeVoiceAgent(_ config:) throws instead of
    // returning a boolean (RunAnywhere+VoiceAgent.swift:24).
    if (arrayBufferToBytes(bytes).byteLength === 0) {
      throw SDKException.componentNotReady(
        'Voice agent initialization returned an empty result'
      );
    }
    logger.info('Voice agent initialized successfully');
  } catch (error) {
    const msg = error instanceof Error ? error.message : String(error);
    logger.error(`Failed to initialize voice agent: ${msg}`);
    throw error;
  }
}

/**
 * Default Silero VAD model id seeded by every example app's catalog.
 * Exposed so callers do not hard-code the string when invoking
 * `ensureDefaultVAD(...)`. Mirrors Swift `RunAnywhere.defaultVADModelID`.
 */
export const defaultVADModelID = 'silero-vad';

/**
 * Ensure a VAD model is loaded in the canonical lifecycle before a voice
 * agent session starts. When no VAD model is currently registered for
 * `MODEL_CATEGORY_VOICE_ACTIVITY_DETECTION`, attempts to load the
 * catalogued default (`defaultVADModelID`, Silero) so the voice agent's
 * speech-start / speech-end events fire. The energy-based fallback does
 * not produce the lifecycle events the voice-agent orchestrator listens
 * for, so without a VAD lifecycle load the session stays silent after
 * init.
 *
 * Idempotent: returns `true` immediately when a VAD model is already
 * loaded. Logs (but does not throw) when the optional auto-load fails;
 * callers may inspect the return value to decide whether to surface a
 * warning. Mirrors Swift `ensureDefaultVAD(modelID:)`.
 *
 * @param modelID VAD model id to auto-load when none is current. When
 *   omitted, falls back to `defaultVADModelID`.
 * @returns `true` when a VAD model is loaded after the call; `false`
 *   when no VAD model is loaded (auto-load failed or skipped).
 */
export async function ensureDefaultVAD(modelID?: string): Promise<boolean> {
  // Swift parity: `guard isInitialized else { return false }`
  // (RunAnywhere+VoiceAgent.swift:58).
  if (!isSDKInitialized()) return false;
  if (!isNativeModuleAvailable()) return false;

  const snapshot = await currentModel(
    CurrentModelRequest.fromPartial({
      category: ModelCategory.MODEL_CATEGORY_VOICE_ACTIVITY_DETECTION,
    })
  );
  if (snapshot && snapshot.found && snapshot.modelId.length > 0) {
    return true;
  }

  const targetID = modelID ?? defaultVADModelID;
  if (targetID.length === 0) return false;

  logger.info(`Auto-loading default VAD '${targetID}' for voice-agent session`);

  const result = await loadModel(
    ModelLoadRequest.fromPartial({
      modelId: targetID,
      category: ModelCategory.MODEL_CATEGORY_VOICE_ACTIVITY_DETECTION,
    })
  );
  if (!result.success) {
    logger.warning(
      `Default VAD '${targetID}' auto-load failed: ${result.errorMessage} — voice agent will use energy fallback`
    );
    return false;
  }
  return true;
}

/**
 * Initialize voice agent using already-loaded models.
 *
 * Composes a `VoiceAgentComposeConfig` from the canonical model lifecycle
 * (`currentModel`) snapshots for `.speechRecognition`, `.language`, and
 * `.speechSynthesis`, then forwards to `voiceAgentInitializeProto`. Mirrors
 * the Swift `initializeVoiceAgentWithLoadedModels(ttsVoiceID:ensureVAD:)` API.
 *
 * When `ensureVAD` is `true` (default), the SDK guarantees that a VAD model
 * is loaded into the canonical lifecycle before initialization runs via
 * {@link ensureDefaultVAD}. Without this the session would silently fall back
 * to the energy-based detector and the C++ voice agent's speech-start /
 * speech-end lifecycle events would not fire. Set to `false` only if the
 * caller has already loaded an explicit VAD model (or knows the energy
 * fallback is acceptable for the deployment).
 *
 * @param ttsVoiceID Optional explicit voice id to pass through to the TTS
 *   engine. For multi-voice TTS engines, this selects the voice within the
 *   loaded model and is semantically distinct from the TTS model id. When
 *   `undefined` (default), the engine's default voice is used. Never reuse
 *   the TTS model id here — model id ≠ voice id.
 * @param ensureVAD Whether to auto-load the catalogued default VAD when no
 *   `MODEL_CATEGORY_VOICE_ACTIVITY_DETECTION` model is loaded. Defaults to
 *   `true`.
 *
 * Mirrors Swift `initializeVoiceAgentWithLoadedModels(ttsVoiceID:ensureVAD:)`.
 */
export async function initializeVoiceAgentWithLoadedModels(
  ttsVoiceID?: string,
  ensureVAD: boolean = true
): Promise<void> {
  // Swift parity: guard isInitialized (RunAnywhere+VoiceAgent.swift:118-120).
  requireInitialized();
  const native = ensureNative();
  // Swift parity: RunAnywhere+VoiceAgent.swift:122 gates on ensureServicesReady.
  await ensureServicesReady();
  try {
    if (ensureVAD) {
      await ensureDefaultVAD();
    }

    const sttSnap = await currentModel(
      CurrentModelRequest.fromPartial({
        category: ModelCategory.MODEL_CATEGORY_SPEECH_RECOGNITION,
      })
    );
    const llmSnap = await currentModel(
      CurrentModelRequest.fromPartial({
        category: ModelCategory.MODEL_CATEGORY_LANGUAGE,
      })
    );
    const ttsSnap = await currentModel(
      CurrentModelRequest.fromPartial({
        category: ModelCategory.MODEL_CATEGORY_SPEECH_SYNTHESIS,
      })
    );

    const missing: string[] = [];
    if (!sttSnap?.found || !sttSnap.modelId) missing.push('STT');
    if (!llmSnap?.found || !llmSnap.modelId) missing.push('LLM');
    if (!ttsSnap?.found || !ttsSnap.modelId) missing.push('TTS');
    if (missing.length > 0) {
      throw SDKException.of(
        ErrorCodeProto.ERROR_CODE_MODEL_NOT_LOADED,
        `Cannot initialize voice agent: Models not loaded: ${missing.join(', ')}`
      );
    }

    logger.info('Initializing voice agent with loaded models...');
    const composeConfig = VoiceAgentComposeConfig.create({
      sttModelId: sttSnap!.modelId,
      llmModelId: llmSnap!.modelId,
      ...(ttsVoiceID && ttsVoiceID.length > 0 ? { ttsVoiceId: ttsVoiceID } : {}),
    });
    await native.voiceAgentInitializeProto(
      encodeProtoMessage(composeConfig, VoiceAgentComposeConfig)
    );
    logger.info('Voice agent initialized with loaded models');
  } catch (error) {
    const msg = error instanceof Error ? error.message : String(error);
    logger.error(`Failed to initialize voice agent: ${msg}`);
    throw error;
  }
}

/**
 * Process a complete voice turn: audio -> transcription -> response -> speech.
 *
 * Matches Swift: `RunAnywhere.processVoiceTurn(_:)`.
 */
export async function processVoiceTurn(
  audioData: ArrayBuffer | Uint8Array
): Promise<VoiceAgentResult> {
  // Swift parity: guard isInitialized (RunAnywhere+VoiceAgent.swift:189-191).
  requireInitialized();
  const native = ensureNative();
  // Swift parity: RunAnywhere+VoiceAgent.swift:182 gates on ensureServicesReady.
  await ensureServicesReady();
  try {
    const resultBytes = await native.voiceAgentProcessTurnProto(
      audioToArrayBuffer(audioData)
    );
    const bytes = arrayBufferToBytes(resultBytes);
    if (bytes.byteLength === 0) {
      throw SDKException.protoDecodeFailed('voiceAgentProcessTurnProto');
    }
    return VoiceAgentResultMessage.decode(bytes);
  } catch (error) {
    const msg = error instanceof Error ? error.message : String(error);
    logger.error(`Voice turn failed: ${msg}`);
    throw error;
  }
}

/**
 * Get the native voice-agent handle.
 *
 * Internal bridge detail for `streamVoiceAgent()`.
 */
async function getVoiceAgentHandle(): Promise<number> {
  const native = ensureNative();
  return native.getVoiceAgentHandle();
}

/** Cleanup voice-agent resources. */
export async function cleanupVoiceAgent(): Promise<void> {
  if (!isNativeModuleAvailable()) return;
  const native = requireNativeModule();
  logger.info('Cleaning up voice agent...');
  await native.cleanupVoiceAgent();
  logger.info('Voice agent cleaned up');
}

/**
 * Stream voice agent events as an AsyncIterable<VoiceEvent>.
 *
 * This is the canonical cross-SDK public method for voice agent streaming.
 * The iterable is constructed synchronously; the native handle is obtained
 * lazily on the first `next()` call, matching Swift's synchronous
 * `AsyncStream` return and the existing RN `generateStream` pattern.
 *
 * Matches Swift: `RunAnywhere.streamVoiceAgent() -> AsyncStream<RAVoiceEvent>`.
 *
 * Usage:
 *   for await (const evt of RunAnywhere.streamVoiceAgent()) { handleEvent(evt) }
 */
export function streamVoiceAgent(): AsyncIterable<VoiceEvent> {
  return {
    async *[Symbol.asyncIterator]() {
      // Swift parity: not-initialized finishes the stream silently
      // (RunAnywhere+VoiceAgent.swift:211-214).
      if (!isSDKInitialized()) {
        return;
      }
      await ensureServicesReady();
      const handle = await getVoiceAgentHandle();
      const adapter = new VoiceAgentStreamAdapter(handle);

      // Swift parity (RunAnywhere+VoiceAgent.swift:225-236): while the event
      // stream is consumed, a platform mic driver captures audio and feeds raw
      // frames to the core via `voiceAgentFeedAudioProto`; the core segments
      // utterances and runs the turn pipeline itself (no SDK-side VAD). The C
      // ABI owns no microphone access — without this driver the pipeline gets
      // no audio buffer and stays "listening" forever (dead-air), per the
      // rac_voice_agent.h Audio-Ingress Contract. Events emitted during each
      // turn fan out to this same handle callback, so collectors see them.
      const micDriver = new VoiceAgentMicDriver();
      void micDriver.start().catch((error) => {
        logger.error(
          `Voice-agent mic driver stopped: ${error instanceof Error ? error.message : String(error)}`
        );
      });
      try {
        yield* adapter.stream();
      } finally {
        // Breaking out of the consuming loop (or unsubscribe) tears down mic
        // capture, mirroring Swift's `defer { micTask.cancel() }`.
        await micDriver.stop();
      }
    },
  };
}
